### PR TITLE
Add controlled decomposition using ZYZ decomposition

### PR DESCRIPTION
**Pull request by mudit2812**
**Context:**
This PR works towards adding decompositions for controlled operations

**Description of the Change:**
Added new file, `pennylane/ops/op_math/controlled_decompositions.py` with function `ctrl_decomp_zyz` such that given a single-qubit target operation and control wires, the function creates a decomposition for the controlled operation using the decomposition shown with Lemma 4.3 and 5.1 of [this paper](https://arxiv.org/pdf/quant-ph/9503016.pdf).

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Comment by mudit2812** [comment]
Testing for `qml.ops.ctrl_decomp_zyz`currently uses a circuit to compare a probability measurement with the decomposed operations to an expected probability measurement. I'm not particularly fond of this way of testing as comparing the decomposed matrix with the expected matrix would be more rigorous, but global phases being added by the decomposition to the target wires is causing tests to fail with matrix comparisons.

Any ideas/feedback is appreciated :)

File: tests/ops/op_math/test_controlled_decompositions.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        #     for decomp_op in decomps
+        # )
+
+        # res = reduce(np.matmul, decomp_mats)
+        # expected = expected_op.matrix()
+        # assert np.allclose(res, expected, atol=tol, rtol=0)
+
+        dev = qml.device("default.qubit", wires=4)
+
+        @qml.qnode(dev)
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_decompositions.py:55]: While these "integration" tests can be important, we want to also have "unit" tests that focus just one the function itself, without depending on things like qnodes, devices, queuing, etc.  Instead, we want to have tests focusing just on "Is this function returning what I think it should?" and "Is what I think it should be returning actually correct?"

For example in testing this function, we could use `qml.equal` to compare each entry to the expected operation.  Though I do think you're right about comparing `matrix()` as well.

**Comment by mudit2812** [comment]
[sc-31253]

**Comment by AlbertMitjans** [comment]
Tagging @KetpuntoG because he was interested in this task

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+from pennylane.operation import Operator
+from pennylane.wires import Wires
+
+
+def ctrl_decomp_zyz(target_operation: Operator, control_wires: Wires):
+    """Decompose a controlled single-qubit operation given the target operation
+
+    This function decomposes the controlled version of a single-qubit operation
+    given a target operation using the decomposition defined in section 5 of
+    `Barenco et al. (1995) <https://arxiv.org/abs/quant-ph/9503016>`_.
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_decompositions.py:28]: Do we need to say it twice? Just a suggestion:

```suggestion
    """This function decomposes the controlled version of a single-qubit operation
    given a target operation using the decomposition defined in section 5 of
    `Barenco et al. (1995) <https://arxiv.org/abs/quant-ph/9503016>`_.
```

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+    .. note:: This method will add a global phase for target operations that do not
+    belong to the SU(2) group.
+
+    Args:
+        target_operation (~.operation.Operator): the target operation to decompose
+        control_wires (~.wires.Wires): the control wire of the operation
```
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_decompositions.py:35]:
```suggestion
        control_wires (~.wires.Wires): the control wires of the operation
```

File: pennylane/ops/op_math/controlled_decompositions.py (hunk 2 above)
Comment (AlbertMitjans) [inline pennylane/ops/op_math/controlled_decompositions.py:28]: Do we need to say it twice? Just a suggestion:

```suggestion
    """This function decomposes the controlled version of a single-qubit operation
    given a target operation using the decomposition defined in section 5 of
    `Barenco et al. (1995) <https://arxiv.org/abs/quant-ph/9503016>`_.
```

**Review by AlbertMitjans — COMMENTED** [review]
Good job! 🚀 💯 

Do we want this as a standalone function, or should we move this implementation inside `Controlled.decomposition`?

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+    Returns:
+        list[Operation]: the decomposed operations
+
+    Raises:
+        ValueError: if `target_operation` is not a single-qubit operation
+
```
Comment (trbromley) [inline pennylane/ops/op_math/controlled_decompositions.py:42]: Please could you add an example usage?

**Comment by mudit2812** [comment]
> Do we want this as a standalone function, or should we move this implementation inside `Controlled.decomposition`?

I don't think this should be moved to `Controlled.decomposition` primarily because that would imply the function only decomposes instances of `qml.Controlled`. I think it might make more sense for `Controlled.decomposition` to use it as a helper function instead.

If the function were to be moved, I would suggest moving it to `pennylane.transforms.decompositions`.

File: pennylane/ops/op_math/controlled_decompositions.py (hunk 2 above)
Comment (mudit2812) [reply pennylane/ops/op_math/controlled_decompositions.py]: I think the first line is important for the docs so that there is a one line description that can be rendered.

File: pennylane/transforms/decompositions/controlled_decompositions.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        ctrl_op = qml.ctrl(op, [0, 1])
+        ctrl_decomp = qml.ops.ctrl_decomp_zyz(op, [0, 1])
+
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def expected_circuit():
+            for i in range(3):
+                qml.Hadamard(i)
+            qml.apply(ctrl_op)
```
Comment (timmysilv) [inline pennylane/transforms/decompositions/controlled_decompositions.py:59]: you can just pass in the op as an argument for `qml.apply` to call so you don't have to define the same QNode twice. same in tests below

**Review by timmysilv — COMMENTED** [review]
sweetness! do you intend on getting to those TODOs in this PR?

File: pennylane/transforms/decompositions/controlled_decompositions.py (hunk 5 above)
Comment (mudit2812) [reply pennylane/transforms/decompositions/controlled_decompositions.py:59]: There isn't much benefit to doing that because the decomposition is a list of `Operator`s so for the 2nd example I need to iterate through that list anyway and in terms of lines of code there will be virtually no difference.

**Comment by mudit2812** [comment]
> sweetness! do you intend on getting to those TODOs in this PR?

I think for now they will have to stay as TODOs because currently getting rid of the global phase requires some really painful decomposition, and the global phase doesn't affect any circuit measurements.

**Review by albi3ro — COMMENTED** [review]
On placing this function in the `transforms` folder:

Historically, we have used `transforms` as kinda "dumping ground" for anything that converts one thing into something else. 

I'd like to move toward grouping source code based on:
* What depends on it
* What it depends on

This will serve to lower the spaghetti of circular dependencies we currently have in Pennylane. It's a hard problem, especially with Python, but we can incrementally get better over time.

In this case, `Controlled.decomposition` will depend on it.

It's intended to be a helper function that `Controlled.decomposition` can choose to use in applicable situations.  Developers and power-users could track this function down and manually use it, but the vast majority of the time this will be used when `Controlled.decomposition` is looking to decompose an operation.

File: pennylane/transforms/decompositions/controlled_decompositions.py [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        )
+
+    target_wire = target_operation.wires
+
+    if len(control_wires) > 2:
+        control_gate = qml.MultiControlledX(wires=control_wires + target_wire)
+    elif len(control_wires) == 2:
+        control_gate = qml.Toffoli(wires=control_wires + target_wire)
+    else:
+        control_gate = qml.CNOT(wires=control_wires + target_wire)
```
Comment (albi3ro) [inline pennylane/transforms/decompositions/controlled_decompositions.py:92]: Given we have:
```
>>> qml.MultiControlledX(wires=(0,1)).decomposition()
[CNOT(wires=[0, 1])]
```

I'm fine with just using `MultiControlledX` for all cases.

File: pennylane/transforms/decompositions/controlled_decompositions.py [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+    try:
+        phi, theta, omega = target_operation.single_qubit_rot_angles()
+    except NotImplementedError:
+        zyz_decomp = qml.transforms.zyz_decomposition(qml.matrix(target_operation), target_wire)[0]
+        phi, theta, omega = zyz_decomp.single_qubit_rot_angles()
+
+    return [
+        qml.RZ(phi, wires=target_wire),
+        qml.RY(theta / 2, wires=target_wire),
+        control_gate,
```
Comment (albi3ro) [inline pennylane/transforms/decompositions/controlled_decompositions.py:103]: Here's a little piece of developer knowledge about an internal peculiarity of pennylane.

`Operator.expand` is constructed by "queuing" `Operator.decomposition`. Basically treating it like a quantum function.

```
with qml.queuing.AnnotatedQueue() as q:
    op.decomposition()
```

Operators are queued when they are initialized.  So if we want an operation to be after another one in the queue, we need to initialize it afterward. This also means that if we want the "same" operation multiple times in the `AnnotatedQueue`, we need to either reinitialize it or use `qml.apply` to copy and queue the original operation.

I'm not a huge fan of this "implicit assumption" we make everywhere and I am pushing to minimize the influence of queuing in PennyLane, but it's a long process.

Long story short, we need to instantiate `control_gate` here and on line 106.

**Comment by josh146** [comment]
> On placing this function in the transforms folder

From what I recall, the main motivation to moving non-trivial generalized decompositions into the `transforms/decompositions` folder was to make them easily importable by end-users, who would use them as part of their [`qml.compile`](https://docs.pennylane.ai/en/stable/code/api/pennylane.compile.html) pipeline.

File: pennylane/transforms/decompositions/controlled_decompositions.py (hunk 7 above)
Comment (timmysilv) [reply pennylane/transforms/decompositions/controlled_decompositions.py:103]: fly-by comment - this planned transition from queuing a decomposition to creating a script from one might make things messy. My gut instinct here was to just use `qml.apply(control_gate)` to fix this issue, but if someone calls this function whilst not in a queuing context, it will raise an error (which I don't think it should)

**Review by timmysilv — COMMENTED** [review]
no worries on dealing with those tasks later. one ask though: rather than leaving a TODO comment in the code, please make an issue/story for yourself and remove the line from source

File: pennylane/transforms/decompositions/controlled_decompositions.py (hunk 7 above)
Comment (mudit2812) [reply pennylane/transforms/decompositions/controlled_decompositions.py:103]: Thanks for the insight! I think the simplest course of action is to replace `control_gate` with `qml.MultiControlledX` as that can be decomposed into `CNOT` or `Toffoli` for 1 or 2 control wires respectively as @albi3ro pointed out. That takes away both of the issues that have been brought up.

**Comment by mudit2812** [comment]
> On placing this function in the `transforms` folder

If the primary use case of the decomposition is as a helper for `qml.Controlled`, then I agree that the function probably belongs in `pennylane/ops/op_math`. As far as the `qml.compile` pipeline goes, it might still make more sense to have the decomposition in `op_math` as a helper function so that a pipeline tape transform relying on operation decomposition can directly use `qml.Controlled.decomposition()` rather than calling `ctrl_decomp_zyz` manually. Let me know what the consensus is and I'll move (or not move) the file back to `op_math` accordingly.

File: tests/transforms/test_decompositions.py [hunk 8]
```diff
@@ -1025,3 +1027,108 @@ def wrapped_decomposition(U):
...
+        res = decomp_circuit()
+        expected = expected_circuit()
+        assert np.allclose(res, expected, atol=tol, rtol=0)
+
+    @pytest.mark.parametrize("op", su2_ops)
+    @pytest.mark.parametrize("control_wires", ([1], [1, 2], [1, 2, 3]))
+    def test_decomposition_matrix(self, op, control_wires, tol):
+        """Tests that the matrix representation of the controlled ZYZ decomposition
+        of a single-qubit operation is correct"""
+        # TODO: Update tests to include non-SU(2) unitaries once supported
```
Comment (timmysilv) [inline tests/transforms/test_decompositions.py:1100]: please open an issue for this if it isn't blocking

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    **Example**
+
+    .. code-block:: python
+
+        # We can create a controlled operation using `qml.ctrl`, or by creating the
+        # decomposed controlled version of using `qml.ctrl_decomp_zyz`.
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_decompositions.py:48]: These comments can just be moved outside of the code block.

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+                qml.Hadamard(i)
+            qml.apply(ctrl_op)
+            return qml.probs()
+
+        @qml.qnode(dev)
+        def decomp_circuit():
+            for i in range(3):
+                qml.Hadamard(i)
+            for op in ctrl_decomp:
+                qml.apply(op)
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_decompositions.py:67]:
```suggestion
            qml.ops.ctrl_decomp_zyz(op, [0,1])
```
Calling the decomposition will queue the operations in the decomposition, do we don't need to pre-initialize and then apply.

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        ctrl_op = qml.ctrl(op, [0, 1])
+        ctrl_decomp = qml.ops.ctrl_decomp_zyz(op, [0, 1])
+
+        dev = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev)
+        def expected_circuit():
+            for i in range(3):
+                qml.Hadamard(i)
+            qml.apply(ctrl_op)
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_decompositions.py:59]:
```suggestion
            qml.ctrl(op, [0,1])
```
Same thing here.  We can just call `ctrl` here instead of pre-initializing and then using `apply`.  This is a little more readable and more in line with how we tend to style things.

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+            for i in range(3):
+                qml.Hadamard(i)
+            for op in ctrl_decomp:
+                qml.apply(op)
+            return qml.probs()
+
+    Measurements on both circuits will give us the same results:
+
+    >>> expected_circuit()
+    tensor([0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125, 0.125], requires_grad=True)
```
Comment (albi3ro) [inline pennylane/ops/op_math/controlled_decompositions.py:72]: This doesn't quite look like a rotation gate has actually been applied, so we can't actually see the effect of the rotation gate.

I think this is because of the `Hadamard` of the target qubit.  `H |0>` is an eigenstate of the `RX` operator, so it's just applying a phase that doesn't show up in the probability.

File: tests/ops/op_math/test_controlled_decompositions.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        """Tests that the matrix representation of the controlled ZYZ decomposition
+        of a single-qubit operation is correct"""
+        decomps = ctrl_decomp_zyz(op, Wires(control_wires))
+        expected_op = qml.ctrl(op, control_wires)
+        decomp_mats = [
+            np.kron(np.eye(2 ** len(control_wires)), decomp_op.matrix())
+            if not isinstance(decomp_op, qml.MultiControlledX)
+            else decomp_op.matrix()
+            for decomp_op in decomps
+        ]
```
Comment (albi3ro) [inline tests/ops/op_math/test_controlled_decompositions.py:101]: You should be able to just do `qml.matrix(ctrl_decomp_zyz)(op, control_wires)`. `ctrl_decomp_zyz` should behave like any other quantum function.

File: pennylane/ops/op_math/controlled_decompositions.py [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+            f"got {target_operation.__class__.__name__}."
+        )
+
+    target_wire = target_operation.wires
+
+    try:
+        phi, theta, omega = target_operation.single_qubit_rot_angles()
+    except NotImplementedError:
+        zyz_decomp = qml.transforms.zyz_decomposition(qml.matrix(target_operation), target_wire)[0]
+        phi, theta, omega = zyz_decomp.single_qubit_rot_angles()
```
[the remark this answers is not in the document]
Comment (timmysilv) [reply pennylane/ops/op_math/controlled_decompositions.py:90]: can we get an integration test where this decomp is called in a recording context (like a QNode) that would have caught this unexpected queuing? oh decompositions...

**Review by timmysilv — COMMENTED** [review]
looks pretty close to ready! just little things left

**Review by timmysilv — APPROVED** [review]
